src/cts/src/HTreeBuilder.cpp

```diff
@@ -363,103 +346,463 @@ void adjustToplevelTopology(Point<double>& a,
-    if (x >= x1 && x <= x2 && y >= y1 && y <= y2) {
-      points.emplace_back(x, y);
+  return branchPoint;
+}
+
+void HTreeBuilder::addCandidatePointsAlongBlockage
+    (Point<double> point,
+     Point<double> parentPoint,
+     double targetDist,
+     int scalingFactor,
+     std::vector<Point<double>>& candidates,
+     std::string location)
+{
+  double x1, y1, x2, y2;
+  if (findBlockage(point, scalingFactor, x1, y1, x2, y2)) {
+    Point<double> point2 = point;
+    double px = parentPoint.getX();
+    double py = parentPoint.getY();
+    // clang-format off
+    debugPrint(logger_, CTS, "legalizer", 3, "  {} corner {} of Manhattan sqaure " 
+	       "is inside blockage ({:0.3f} {:0.3f}) ({:0.3f} {:0.3f})", location, point, x1, y1, x2, y2);
+    // clang-format on
+    if (location == "top") {
+      //        ----------
+      //        |        |
+      //        |   top  |
+      //        |   / \  |
+      // (x1, y1)--x---x-(x2, y1)
+      //          /  .
+      //          (px, py)
+      //
+      // (px - x) + (y1 - py) = targetDist
+      // (x - px) + (y1 - py) = targetDist
+      point2.setX(-targetDist - py + px + y1);
+      point2.setY(y1);
+      candidates.emplace_back(point2);
+      point2.setY(targetDist + py + px - y1);
+      candidates.emplace_back(point2);
+    } else if (location == "right") {
+      // (x1, y2)---------
+      //       \ |        |
+      //         x        |
+      //         |\       |
+      //   .     | right  |
+      // (px,py)| /       |
+      //         x        |
+      // (x1, y1)---------
+      // (x1 - px) + (y - py) = targetDist
+      // (x1 - px) + (py - y) = targetDist
+      point2.setX(x1);
+      point2.setY(targetDist + py + px - x1);
+      candidates.emplace_back(point2);
+      point2.setY(-targetDist + py - px + x1);
+      candidates.emplace_back(point2);
+    } else if (location == "bottom") {
+      point2.setX(-targetDist + py + px - y2);
+      point2.setY(y2);
+      candidates.emplace_back(point2);
+      point2.setX(targetDist - py + px + y2);
+      candidates.emplace_back(point2);
+    } else {
+      assert(location == "left");
+      point2.setX(x2);
+      point2.setY(targetDist + py - px + x2);
+      candidates.emplace_back(point2);
+      point2.setY(-targetDist + py + px - x2);
+      candidates.emplace_back(point2);
     }
   }
-  if (points.empty()) {
-    points.emplace_back(px - leng, py);
-    points.emplace_back(px + leng, py);
-    points.emplace_back(px, py + leng);
-    points.emplace_back(px, py - leng);
-
-    const double leng2 = leng / 2.0;
-    points.emplace_back(px - leng2, py + leng2);
-    points.emplace_back(px + leng2, py + leng2);
-    points.emplace_back(px - leng2, py - leng2);
-    points.emplace_back(px + leng2, py - leng2);
+}
+
+void HTreeBuilder::checkLegalityAndCost(Point<double> oldLoc,
+                                        Point<double> newLoc,
+                                        Point<double> parentPoint,
```

Comment:
Same from before, could these points be const&?

Reply:
Yes, done.